### PR TITLE
python 3.10.0

### PR DESCRIPTION
**Pull request by Moisan**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

First try with Python 3.10.0 release candidate 1. https://www.python.org/downloads/release/python-3100rc1/

Audit currently fails with 
```
  * Versioned formulae in homebrew/core should use `keg_only :versioned_formula`
  * Stable version URLs should not contain rc1
```

Note that, compared to `python@3.9`, I removed the `decimal` diff since I couldn't find the one for for Python 3.10. 

Release candidate 2 is scheduled for Monday, 2021-09-06.

### CODE REVIEW

**Comment by branchvincent** [comment]
i believe both audits can be ignored by updating [versioned_keg_only_allowlist](https://github.com/Homebrew/homebrew-core/blob/master/audit_exceptions/versioned_keg_only_allowlist.json) and [unstable_allowlist](https://github.com/Homebrew/homebrew-core/blob/master/audit_exceptions/unstable_allowlist.json)

**Comment by cho-m** [comment]
Looks like release candidate successfully builds, but CI probably won't find any dependents due to versioned `depends_on "python@3.9"` (~582 formulae).

**Comment by iMichka** [comment]
Yes. If we want to test this correctly, we need to Rev bump all these 582 formulae and update the Python 3.9 dependency to 3.10. In 582 separate commits.

**Comment by Moisan** [comment]
👍 I'll first try all formulae that are in `brew uses python@3.9` and we'll see how it goes.

**Comment by cho-m** [comment]
Sounds good as remaining will require `CI-build-dependents-from-source` to correctly test.

May need to use Self-Hosted Linux given how long Python 3.9.7 CI took.

**Comment by Moisan** [comment]
A few issues that I spotted based on the current ARM failures (138 fails):
* Python 3.10 [removes `bdist_wininst`](https://github.com/python/cpython/commit/0e2a0f72cc9af0899eacb5604e44a563c0b06110). `setup_tools` [removed it in version `v55.0.0`](https://github.com/pypa/setuptools/blob/7685e2393c9f44bcff13d62361f2355595d856ba/CHANGES.rst#v5500). Some formulae that were bumped in this PR uses a version of `setuptools` resource that is older than that, which leads to build failure.
* `poetry` next release [should support Python 3.10](https://github.com/python-poetry/poetry/issues/3187)
* `sphinx-doc` 4.2.0 (next release) [should support Python 3.10](https://github.com/sphinx-doc/sphinx/issues/9562)

**Comment by branchvincent** [comment]
`sphinx-doc` fix landed with https://github.com/Homebrew/homebrew-core/pull/85095

**Comment by Moisan** [comment]
I rebased on master and updated to Python 3.10rc2.

**Comment by branchvincent** [comment]
`sip` just released [3.10 support](https://www.riverbankcomputing.com/news/SIP_v6.2.0_Released), landed in https://github.com/Homebrew/homebrew-core/pull/86393

**Comment by Moisan** [comment]
Python 3.10 should officially be released today. My plan is to follow the [previous `python@3.9` migration](https://github.com/Homebrew/homebrew-core/issues/62201) pattern:
* Merge `python@3.10` and  `python-tk@3.10` within this PR
* Open individual PR for independent formulae first: depend on `python@3.9` only directly and have no reverse dependencies. I can probably write a script to do that automatically.
* Migrate other formulae

**Comment by cclauss** [comment]
Python 3.10.0 binaries are now available at https://www.python.org/ftp/python/3.10.0/

File: Formula/python@3.10.rb [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+  link_overwrite "lib/pkgconfig/python3.pc"
+  link_overwrite "lib/pkgconfig/python3-embed.pc"
+  link_overwrite "Frameworks/Python.framework/Headers"
+  link_overwrite "Frameworks/Python.framework/Python"
+  link_overwrite "Frameworks/Python.framework/Resources"
+  link_overwrite "Frameworks/Python.framework/Versions/Current"
+
+  resource "setuptools" do
+    url "https://files.pythonhosted.org/packages/db/e2/c0ced9ccffb61432305665c22842ea120c0f649eec47ecf2a45c596707c4/setuptools-57.4.0.tar.gz"
+    sha256 "6bac238ffdf24e8806c61440e755192470352850f3419a52f26ffe0a1a64f465"
```
Comment (branchvincent) [inline Formula/python@3.10.rb:54]:
```suggestion
    url "https://files.pythonhosted.org/packages/1e/5c/3d7b3d91a86d71faf5038c5d259ed36b5d05b7804648e2c43251d574a6e6/setuptools-58.2.0.tar.gz"
    sha256 "2c55bdb85d5bb460bd2e3b12052b677879cffcf46c0c688f2e5bf51d36001145"
```

**Review by branchv — DISMISSED** [review]
`match` statement, here we come!

**Review by branchv — APPROVED** [review]
oops, forgot `--no-autosquash`

**Comment by carlocab** [comment]
Have we made `python@3.9` keg-only?

**Comment by branchvincent** [comment]
i guess we need to migrate the aliases as well as make `python@3.9` key-only. let me do that

**Pull request closed without merging**